pkg/controllers/nodeoverlay/validation/store.go

```diff
@@ -0,0 +1,264 @@
+func (s *InstanceTypeStore) UpdateInstanceTypeCapacity(nodePoolName string, instanceTypeName string, nodeOverlay v1alpha1.NodeOverlay, resourceToUpdate corev1.ResourceList) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	_, foundNodePool := s.updates[nodePoolName]
+	if !foundNodePool {
+		return
+	}
+	_, foundInstanceTypeUpdate := s.updates[nodePoolName][instanceTypeName]
+	if !foundInstanceTypeUpdate {
+		return
+	}
+
+	s.updates[nodePoolName][instanceTypeName].Capacity = append(s.updates[nodePoolName][instanceTypeName].Capacity, CapacityUpdate{
+		overlayName:   nodeOverlay.Name,
+		OverlayUpdate: resourceToUpdate,
+		weight:        nodeOverlay.Spec.Weight,
+	})
+}
+
+// UpdateInstanceTypeOffering add a new Price overlay update to the associated instance type.
+// This add an Price update to the store once after it has been validated.
+func (s *InstanceTypeStore) UpdateInstanceTypeOffering(nodePoolName string, instanceTypeName string, nodeOverlay v1alpha1.NodeOverlay, offerings cloudprovider.Offerings) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	_, foundNodePool := s.updates[nodePoolName]
+	if !foundNodePool {
+		return
+	}
+	_, foundInstanceTypeUpdate := s.updates[nodePoolName][instanceTypeName]
+	if !foundInstanceTypeUpdate {
+		return
+	}
+
+	overlayPriceChange := lo.Ternary(nodeOverlay.Spec.Price == nil, nodeOverlay.Spec.PriceAdjustment, nodeOverlay.Spec.Price)
+	s.updates[nodePoolName][instanceTypeName].Price = append(s.updates[nodePoolName][instanceTypeName].Price, PriceUpdate{
+		overlayName:   nodeOverlay.Name,
+		OverlayUpdate: overlayPriceChange,
+		Offerings:     offerings,
+		weight:        nodeOverlay.Spec.Weight,
+	})
+}
+
+func (s *InstanceTypeStore) RemoveAllInstanceTypeUpdateFromOverlay(nodeOverlayName string) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	for nodePoolName := range s.updates {
+		for instanceTypeNames := range s.updates[nodePoolName] {
+			s.updates[nodePoolName][instanceTypeNames].Price = lo.Filter(s.updates[nodePoolName][instanceTypeNames].Price, func(update PriceUpdate, _ int) bool {
+				return update.overlayName != nodeOverlayName
+			})
+			s.updates[nodePoolName][instanceTypeNames].Capacity = lo.Filter(s.updates[nodePoolName][instanceTypeNames].Capacity, func(update CapacityUpdate, _ int) bool {
+				return update.overlayName != nodeOverlayName
+			})
+		}
+	}
+}
+
+func (s *InstanceTypeStore) ApplyOverlayOnInstanceTypes(nodePoolName string, its []*cloudprovider.InstanceType) []*cloudprovider.InstanceType {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	result := []*cloudprovider.InstanceType{}
+
+	_, foundNodePoolUpdates := s.updates[nodePoolName]
+	if !foundNodePoolUpdates {
+		return its
+	}
+
+	for _, it := range its {
+		instanceTypeUpdate, foundInstanceTypeUpdate := s.FindInstanceTypesUpdate(nodePoolName, it)
+		if !foundInstanceTypeUpdate {
+			result = append(result, it)
+			continue
+		}
+
+		overriddenInstanceType := &cloudprovider.InstanceType{
```

Comment:
We're already doing a deep-copy of all of the instance types as part of the decorator, is there any reason to do a deep copy again here rather than mutate them in place?

Reply:
This function is intended to be used independent of the decorator, as cloud provider are expected to apply overlays using this function as well. For that reason the deep copy in the decorator was dropped. This also has the benefits that only instance types are overlaid will be copied